src/services/users.js

```diff
@@ -0,0 +1,119 @@
+import { pick } from "lodash";
+
+import API from "../api/users";
+import { UserModel, PasswordModel } from "../models";
+import { DEFAULT_PWD } from "../config";
+import { pwdEncrypt } from "../lib/utils";
+
+export class Service extends API {
+  /**
+   * Ability to inject some middlewares
+   *
+   * @param {string} operation name of operation
+   * @returns {function[]} middlewares
+   */
+  middlewares(operation) {
+    const load = async (ctx, next) => {
+      const { id } = ctx.params;
+      if (id) {
+        ctx.state.user = await UserModel.get(id);
+        if (!ctx.state.user) {
+          throw ctx.throw(404, `user ${id} not found.`);
+        }
+      }
+      await next();
+    };
+    return [load];
+  }
+
+  /**
+   * List all users
+   *
+   * @param {ListUsersRequest} req listUsers request
+   * @param {import("koa").Context} ctx koa context
+   * @returns {ListUsersResponse} A paged array of users
+   */
+  async listUsers(req, ctx) {
+    const { limit = 10, offset = 0, sort, filter = {} } = req.query;
+
+    const result = await UserModel.list({ limit, offset, sort, filter });
+
+    return {
+      body: result.docs,
+      headers: {
+        xTotalCount: result.total,
+      },
+    };
+  }
+
+  /**
+   * Create a user
+   *
+   * @param {CreateUsersRequest} req createUsers request
+   * @param {import("koa").Context} ctx koa context
+   * @returns {CreateUsersResponse} The User created
+   */
+  async createUser(req, ctx) {
+    const { body } = req;
+    const { name } = body;
+    if (!name) {
+      throw ctx.throw(404, "user name not found.");
+    }
+    const repeatUser = await UserModel.findOne({ name });
+
+    if (repeatUser) {
+      if (ctx) throw ctx.throw(404, "user already exsists");
+      else return;
+    }
+    const user = await UserModel.create(body);
+    await PasswordModel.create({

```

Comment:
As there are 2 creating operations it's possible to have inconsistent data around the user.id(e.g. the `PasswordModel.create` fails.
Do you have any idea to fix or mitigate the issue?
You don't need to update the code just put your ideas here.

Reply:
I didn't have concern here before, since mongoosedocument says model.create() returns a promise.

If we try to avoid this problem, I will try :-)
Method #1:
use `new Model()`, it should return user.id, too.
```ES6
    const user = new UserModel(body)
    await user.save()
    await PasswordModel.create({
      user: user.id,
      password: pwdEncrypt(DEFAULT_PWD),
    });
```

Method #2: 
We can put password into User Model as an attribute with`hideJSON: true`, and get rid of Password Model.
Then we only need create User once, without worry about the password creation here.

```ES6 
export const userSchema = new mongoose.Schema(
  {
    name: String,
    _password: { type: String, hideJSON: true }
  },
  {...}
);
```

And we write set and checkpassword method for password:
```ES6
class User {
  // other codes
set password(pwd) {
    this._password = encrypt(pwd)
  }

  checkPassword(pwd) {
    return verify(pwd,this._password)
  }
}
```

Reply:
`hideJSON` is one of property params of `mongoose-hidden` library which is a mongose schema plugin.

https://github.com/mblarsen/mongoose-hidden

Reply:
Thanks for the ideas, but I don't think a promise will always be resolved, it could fail and throw, no?

> I didn't have concern here before, since mongoosedocument says model.create() returns a promise.

Reply:
Sure, when the database got trouble like too much work flood in, it could happen. I will keep that in my mind for the future design.